crates/artifact/src/s3.rs

```diff
@@ -23,11 +23,26 @@ lazy_static! {
 }
 
 const CHUNK_SIZE: usize = 32 * 1024 * 1024;
+const MAX_RETRY_COUNT: u32 = 7;
+
+pub enum S3DownloadMode {
+    AwsSdk,
+    REST,
+}
+const DEFAULT_S3_DOWNLOAD_MODE: S3DownloadMode = S3DownloadMode::REST;
+
+pub trait S3DownloadClient: Send + Sync {
+    fn get_object_size(&self, bucket: String, key: String) -> Result<i64>;
+    fn read_byte_range(&self, bucket: String, key: String, start: i64, end: i64)
```

Comment:
might be slightly more performant to pass in &str for bucket since a String is not always needed